### DISCUSSION TITLE
using web3 client to interact with a smart contract?

### DISCUSSION THREAD

**Discussion by scutdk** [gochain/web3#207] [General]
Hi, I am an experienced CLI user. And I find the web3 CLI wallet is very cool and suit well for my needs.
I've got to understand how to set up and transfer both local tokens and ERC20 tokens to another address.

My next question is how can I interact with a smart contract as a user rather than developer? 
I tried to find any answer from the [README.md](https://github.com/gochain/web3) with no luck.

There're some topics relating to write contract which I am not very sure.
Can you give some details on interacting with smart contracts , e,g, to swap GO for USDC from goswap?

I think their are two steps, approve and make the swap, right?
I don't have any abi file as I am a normal user not the developer.
 How Can I do that with web3 CLI client. Thanks.

**Comment by treeder** [comment]
You need the abi to interact with a contract, to swap on GoSwap, you'll probably want the router contract so you can use this flag: `--abi https://raw.githubusercontent.com/goswap/uniswap-v2-periphery/master/contracts/UniswapV2Router02.abi`

The function is one of the swap ones like: [swapExactTokensForTokens](https://uniswap.org/docs/v2/smart-contracts/router02/#swapexacttokensfortokens)

For approvals, that's part of the token itself which is an ERC20/GO20 contract, so you can use `--abi https://raw.githubusercontent.com/gochain/web3/master/assets/erc20.abi` for the approve call.

**Reply by scutdk** [reply]
I've got some understanding of what an abi file is.
I've also went through some other web3 clients like web3.js and web3.py.
web3 go client is more convenient as it is a shell script cmd.

Is it possible to utilize sendtransaction function as I need to send hex data directly with web3 go client?
The contract is not open sourced, thus I can't get the abi file from etherescan.
I can get the hex data sent to the contract with etherscan.
I've searched the source code of web3 go client and it seems not possible to send hex data directly for now?

**Reply by treeder** [reply]
Doesn't look like `call` has a `data` parameter, but probably wouldn't be too hard to add. It's already in the `replace` command here: https://github.com/gochain/web3/blob/master/cmd/web3/main.go#L279

**Reply by 621625** [reply]
Following

**Comment by treeder** [comment]
Potential PR: https://github.com/gochain/web3/pull/217

**Comment by treeder** [comment]
Upgrade your `web3` and give it a try now that #217 is merged. use `--data` flag.

**Reply by scutdk** [reply]
I've tested the latest version and it seems I still need an abi file when calling with --data flag. The abi file is not accessible in some cases as it is not open sourced.

**Reply by treeder** [reply]
Doh. https://github.com/gochain/web3/issues/226

**Reply by treeder** [reply]
Try now with latest build.

**Reply by scutdk** [reply]
> Try now with latest build.

It's working as designed properly. Thanks for quick update.

**Reply by 621625** [reply]
@scatdk I would like to test this too. Could you perhaps paste the input you use to make the trade here?

**Reply by scutdk** [reply]
> @scatdk I would like to test this too. Could you perhaps paste the input you use to make the trade here?

just copy&paste whatever data you grab from etherscan/bscscan and append to the data field. web3 c call --address <contract_address> --data <input_data_from_etherscan> .

**Reply by 621625** [reply]
I tried that but the value is then incorrect. I. e. even if the original transaction had 1 Matic value the duplicate transaction will have 0 Matic value.

**Reply by treeder** [reply]
related: #232
